design/encryption-spi.md

```diff
@@ -0,0 +1,166 @@
+If the secret is read from the database in encrypted form then the secret is decrypted with the provider before returning it.
+If the provider is not available then error is logged and exception is thrown.
+
+The `encrypted-data` is stored in base64 encoding.
+The data itself is opaque for Keycloak, including the encryption algorithm used for encryption and any possible metadata included in `encrypted-data`.
+These are managed internally by the SPI provider.
+
+> **TODO**: Q: Is lazy encryption acceptable as an approach?
+>
+> Lazy encryption limits the impact to code that handles secrets.
+> The problem is limited to providing and using encrypt/decrypt operations.
+>
+> It may be possible that this approach is too trivial and more complex solution is needed.
+> As an example, consider following use cases:
+>
+> (1) Administrator might want to encrypt all data at once when taking the SPI into use in existing installation.
+>
+> (2) Administrator might want to re-encrypt all data with new encryption key after key rotation.
+>
+> (3) Administrator might want to query for the status of the encryption: are all data encrypted, which encryption provider was used, which encryption keys are in use.
+>
+> (4) Administrator might want to decrypt all data.
+> This could be used to migrate data back to plaintext or to move to another encryption provider.
+> It could be used to recover from unseen issues.
+>
+> Supporting the above use cases would require much wider changes than presented in this document.
+> For example, there could be a "framework" to support marking which data fields are sensitive, making it possible to query and iterate over all sensitive fields in the database.
+
+### Interface
+
+```java
+public interface EncryptionServiceProvider {
+    Optional<byte[]> encrypt(byte[] plaintext);
+    Optional<byte[]> decrypt(byte[] ciphertext);
+}
+```
+
+### Supported fields
+
+Following fields are proposed to be encrypted by the SPI.
+Fields can be included incrementally in the future, and not all fields need to be covered at once.
+
+* Client secret generated by Keycloak.
+* Client secret configured by administrator for identity brokering.
+* External tokens stored in the database by Keycloak when "Store tokens" is configured for identity brokering.
+* Realm keys generated by Keycloak.
+* OTP shared secrets generated by Keycloak.
+* LDAP credentials configured by administrator.
+* SMTP credentials configured by administrator.
+
+There is some overlap with the Vault SPI.
+It is not anticipated that both the Vault SPI and Encryption SPI will be used simultaneously.
+The justification for the overlap are discussed in the Motivation section.
+
+Encryption SPI cannot be used for fields that need to remain searchable in the database.
+This can include fields like username, email, or other fields that can be considered as privacy-sensitive but need to be searchable.
+
+
+### Performance
```

Comment:
I very much doubt this would scale well for things like client secrets and OTP secrets. 

For client secrets hashing may be a better option than storing in an external vault.

Reply:
You are probably right that the performance would not be sufficient.

I came across the [discussion](https://github.com/keycloak/keycloak/discussions/8455) you've had regarding hashing and it seems like a good option as well. My concern is that it only focuses on client secrets, and using encryption might allow us to avoid the backward compatibility issues caused by hashing.

I’ve been studying Kubernetes' KMSv2 implementation for securing `Secrets`. They've managed to create a highly efficient solution by minimizing calls to external encryption services, though it adds complexity on the Kubernetes side rather than in the KMS side. At a high level:

* Each secret is encrypted locally using a data encryption key (DEK).
* A new DEK is generated locally with every secret write.
* DEK generation is based on a key derivation algorithm that uses a seed.
* The seed is encrypted by the remote encryption service.

I believe this approach could work for us as well, and I can present a similar strategy.

Reply:
Hashing does provide an extra layer of security in the fact that the original secret for a client can not be retrieved again through REST APIs. In general I think any secrets stored in a vault (or hashed or whatever) should not be readable through REST APIs.